src/wayland/session_lock/lock.rs

```diff
@@ -0,0 +1,134 @@
+//! ext-session-lock lock.
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+
+use crate::wayland::compositor;
+use crate::wayland::compositor::SurfaceAttributes;
+use _session_lock::ext_session_lock_surface_v1::ExtSessionLockSurfaceV1;
+use _session_lock::ext_session_lock_v1::{Error, ExtSessionLockV1, Request};
+use wayland_protocols::ext::session_lock::v1::server::{self as _session_lock, ext_session_lock_surface_v1};
+use wayland_server::{Client, DataInit, Dispatch, DisplayHandle, Resource};
+
+use crate::wayland::session_lock::surface::{ExtLockSurfaceUserData, LockSurface, LockSurfaceAttributes};
+use crate::wayland::session_lock::{SessionLockHandler, SessionLockManagerState};
+
+/// Surface role for ext-session-lock surfaces.
+const LOCK_SURFACE_ROLE: &str = "ext_session_lock_surface_v1";
+
+/// [`ExtSessionLockV1`] state.
+#[derive(Debug)]
+pub struct SessionLockState {
+    pub(crate) lock_status: Arc<AtomicBool>,
+}
+
+impl SessionLockState {
+    pub(crate) fn new() -> Self {
+        Self {
+            lock_status: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<D> Dispatch<ExtSessionLockV1, SessionLockState, D> for SessionLockManagerState
+where
+    D: Dispatch<ExtSessionLockV1, SessionLockState>,
+    D: Dispatch<ExtSessionLockSurfaceV1, ExtLockSurfaceUserData>,
+    D: SessionLockHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        lock: &ExtSessionLockV1,
+        request: Request,
+        data: &SessionLockState,
+        _display: &DisplayHandle,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            Request::GetLockSurface { id, surface, output } => {
+                // Assign surface a role and ensure it never had one before.
+                if compositor::get_role(&surface).is_some()
+                    || compositor::give_role(&surface, LOCK_SURFACE_ROLE).is_err()
```

Comment:
Even though the protocol says "If the surface already has a role", there is reason to allow clients reusing wl_surfaces, that previously have had the `LOCK_SURFACE_ROLE`.

Other protocols have the same wording, still clients expect to be able to re-use wl_surfaces for XdgSurfaces or LayerSurfaces. So imo the first condition here should be removed.